src/core/api/heap_api.cpp

```diff
@@ -63,6 +63,17 @@ void otHeapFree(void *aPointer)
 
 #else // OPENTHREAD_RADIO
 
+#if OPENTHREAD_CONFIG_EXTERNAL_HEAP_ENABLE
+void *otHeapCAlloc(size_t aCount, size_t aSize)
+{
+    return ot::Instance::HeapCAlloc(aCount, aSize);
+}
+
+void otHeapFree(void *aPointer)
```

Comment:
I think at some later point we may want to remove these public APIs. I see the `TODO` in the header files for these and they are marked to be deprecated.

Reply:
> The challenge/reason may have been due to way the public OT API otHeapCAlloc() and otHeapFree() are defined. They don't have otInstacne *aInstance as parameter, so may not know which instance to use in the multi-instance case (but I guess we can either change them or even better remove them since they are marked to be deprecated).

Those APIs are currently used by nordic boards to setup mbedtls and I think the function signature can not be changed.

Reply:
As mentioned in https://github.com/openthread/openthread/pull/6123#issuecomment-769581199, the function signature is defined by mbedTLS.

https://github.com/openthread/openthread/blob/46f796d4f041c80c716b1ede81eba13c9b9e4f06/third_party/mbedtls/repo/include/mbedtls/platform.h#L130-L140

Reply:
I think the the `otHeapCAlloc` and `otHeapFree` APIs cannot support `MULTIPLE_INSTANCE` unless we enforce `EXTERNAL_HEAP` for `MULTIPLE_INSTANCE`.

I think the intention for those two API is for the application to use OT's internal heap to setup mbedtls. If external heap is enabled, the application should just use `calloc` directly.

We should make it clear that the `otHeapCAlloc` and `otHeapFree` APIs are not available for `MULTIPLE_INSTANCE`. thoughts?

Reply:
I think multiple instance requires external heap due to mbedTLS constraints.